### ISSUE TITLE
'Transactional' package extraction

### ISSUE THREAD

**Issue by maraf** [gitextensions/gitextensions.pluginmanager#8]
Package extraction include write to packages.config should succeed as an unit or any partial outputs should be deleted in case of exception.

Toughts:
Use kind of a transaction log.

**Installation**
- Save package id and version.
- After each file extraction, save a path to log.
- Write entry to packages.config and delete transaction log.

**Unstallation**
- Save package id and temp directory patb.
- Copy each file to temp directory.
- Remove entry from packages.config.
- Delete temp directory.

Check transaction log on PackageManager start.

**Comment by mast-eu** [comment]
Could you please provide more context? How many write accesses are performed to `packages.config`?
How much of a risk do you see to end up with an inconsistent `packages.config`?

**Comment by maraf** [comment]
Every installation and update do two things
- write to packages.config,
- extract content from package. Most packages right now contain only one file, but packages like PM contains more files.

There is always chance that something goes wrong. Now we have no way to check. 
The worst scenario is GE will be crashing all over again, because some dll in user plugins will be missing,.

**Comment by mast-eu** [comment]
> There is always chance that something goes wrong. Now we have no way to check.

You're right, many things can go wrong when dealing with I/O. However, I'm strongly inclined to take the risk for v.1.0.0.
In the worst case, a few users will have to manually delete their `UserPlugins` directories (except PluginManager, of course 😏 ) and install their plugins again.

Do you agree?

**Comment by maraf** [comment]
Absolutely.
